netdisco/xiaomi_gw.py

```diff
@@ -0,0 +1,68 @@
+"""Xiaomi Mi Home (aka Lumi) Gateway discovery."""
+import socket
+import json
+from datetime import timedelta
+
+
+DISCOVERY_PORT = 4321
+DISCOVERY_ADDRESS = "224.0.0.50"
+DISCOVERY_PAYLOAD = b'{"cmd":"whois"}'
+DISCOVERY_TIMEOUT = timedelta(seconds=3)
+
+
+class XiaomiGw(object):
+    """Base class to discover Xiaomi Gateway devices."""
+
+    def __init__(self):
+        """Initialize the discovery."""
+        self.entries = []
+
+    def scan(self):
+        """Scan the network."""
+        self.update()
+
+    def all(self):
+        """Scan and return all found entries."""
+        self.scan()
+        return self.entries
+
+    def update(self):
+        """Scan network for devices."""
+        entries = []
+
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        sock.settimeout(DISCOVERY_TIMEOUT.seconds)
+        sock.sendto(DISCOVERY_PAYLOAD, (DISCOVERY_ADDRESS, DISCOVERY_PORT))
+
+        while True:
+            try:
+                data, (address, _) = sock.recvfrom(1024)
+                print(data, address)
+                try:
+                    entry = json.loads(data.decode("utf-8"))
+                except:  # pylint: disable=bare-except
+                    continue
+                if entry.get("model") == "gateway":
```

Comment:
I think that we should not filter entries, instead return all discovered Xiaomi devices.

Reply:
Hmm, there're few issues with that.

1. This implements discovery specifically for https://home-assistant.io/components/xiaomi/ , which is the Gateway device. The naming in HA is "not ideal", per p.2. So, the above code statement isn't really "filtering", but a simple check that what was discovered is actually a specific device and not something else (random UDP server running on that port).

2. "All Xiaomi devices" are 20-30 *absolutely different* devices nowadays, and might go 200-300 in the future, unless Xiaomi goes bankrupt, which doesn't seem to be a problem so far, given their marketing know-how of selling their name to random Chinese companies to help them sell their stuff to Westeners. I really wouldn't like HA try to do something about Xiaomi car simply because it can work with Xiaomi Gateway (this includes mis-discovering a car as a gateway).

Reply:
I think in that case the filtering should happen in the discoverables. So we can discover a gateway as a separate model. That way if we want to discover something else, we can re-use the discovery code.

Reply:
Ok, I was thinking about this too, and some clarification: discovery protocol implemented by "xiaomi_gw.py" is **not** a discovery protocol for "Xiaomi products", because majority of "Xiaomi products" are made by varying "subsidiaries", which have little in common on technical side (except for integration into Xiaomi's cloud). In particular, the protocol used by "xiaomi_gw.py" is currently known to be used only by Xiaomi Mi Home Gateway. That's why the protocol module is called "xiaomi_gw.py" (discoverable module is called "xiaomi.py", because it should match HA's component (or service?) name, and unfortunately, HA uses "xiaomi" for the *Gateway* name, which will lead to confusion as more wide-varying "Xiaomi" devices are added to HA).

It's possible that this protocol will be used in other "Xiaomi" products, but it will never be the only protocol to discover all "Xiaomi" devices. Hence, I err on the side of specificity, to avoid even remotely the situation that somebody's Xiaomi Water Tap (for example) will be recognized as Home Gateway.

@balloob, you suggest to err on the side of generality. I usually love that, but not this time, due to the reasons above. However, seeing your point, I may imagine following change:

1. xiaomi_gw.py gets renamed to xiaomi_whoami.py, with docstrings that this module implements Xiaomi UDP, JSON based "whoami" protocol, which is currently known to be used by just Gateway.
2. xiaomi.py discoverable filters just "gateway" entries, as this discoverable indeed represent only gateway.

Does that sound good?

Reply:
Sounds good 👍